### DISCUSSION TITLE
Strange behavior/performance difference with two loops (incl. Sources + Testcase)

### DISCUSSION THREAD

**Discussion by 2jfw** [airsdk/Adobe-Runtime-Support#2587] [Q&A] [answered]
I have a scenario which I don't really understand - basically it is two for-loops of the same kind being run and benchmarked.
The first loop finishes way slower than the second one.

E.g.:
time: **8151** ms
vs.
time: **985** ms

I made sure that both loops have no other differences than the variable names.
Below is the code and attached is the test app (50.2.1.1) including a release build (Windows) for this.

My question @ajwfrost: Do you have any idea where the difference comes from?

Sources:
[LoopPerformance.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/11212058/LoopPerformance.zip)

```
package
{

	import flash.display.Sprite;
	import flash.text.TextField;
	import flash.utils.getTimer;
	import flash.utils.setTimeout;


	public class LoopPerformance extends Sprite
	{

		private var _outputTextField : TextField;
		private var _obj : Object  = {};
		private var _obj2 : Object = {};


		public function LoopPerformance()
		{
			createOutputTextField();

			setTimeout(execute,
			           2000); // intentionally delayed call to have the window etc get initialized (/ possible AIR internals e.g. ?)
		}


		private function createOutputTextField() : void
		{
			_outputTextField        = new TextField();
			_outputTextField.width  = 400;
			_outputTextField.height = 300;

			addChild(_outputTextField);
		}


		public function execute() : void
		{
			// Declaring all at the beginning and intentionally not recycling the variables!

			var i : int;
			var j : int;
			var len : int  = 1000000;
			var len2 : int = 1000000;
			var time : Number;
			var time2 : Number;


			// Actual start of test
			time = getTimer();
			for (i = 0; i < len; i++)
			{
				_obj["property" + i] = i;
			}

			_outputTextField.text += "time: " + (getTimer() - time) + "\n";


			time2 = getTimer();
			for (j = 0; j < len2; j++)
			{
				_obj2["property" + j] = j;
			}

			_outputTextField.text += "time: " + (getTimer() - time2) + "\n";
		}
	}
}

```

**Comment by teotigraphix** [comment]
Naively.. I would throw out maybe there is initialization costs on an `Object`(since it's not an Array) when it's key properties are initially set, then once they exist, it's fast access. Just a thought, curious as the what the VM is doing. (maybe hashing the keys)

**Reply by 2jfw** [reply]
> Naively.. I would throw out maybe there is initialization costs on an `Object`(since it's not an Array) when it's key properties are initially set, then once they exist, it's fast access. Just a thought, curious as the what the VM is doing. (maybe hashing the keys)

The thing is, that both loops access two different objects, _obj_ and _obj2_, so both should have the same time getting initialized I think..

**Reply by teotigraphix** [reply]
Oh man, too early haha, yeah wow my bad!

**Reply by 2jfw** [reply]
No problem :)

**Comment by 2jfw** [comment]
Calling the function again a second time does actually result in the nearly same/equal timings:

8479ms (_obj loop)
1469ms (_obj2 loop)
1155ms (_obj loop, 2nd execute call)
1122ms (_obj2 loop, 2nd execute call)

So it seems that some "things" (/internals) seem to get created/initialized in AVM and then can be accessed faster on the second loop (_obj2) in the first iteration or after the second call of `execute` (kind of similar to what you suggested @teotigraphix)

**Reply by 2jfw** [reply]
Maybe it's the _JIT_ compilation and at a later execution of a same or similar construct it uses some kind of cache?

**Reply by ajwfrost** [reply, marked as answer]
I think the "things" are the strings... when you use a string to access a property of an object, it gets internalized, so essentially you've got a million different string concatenations going on, and then each string is set up in the internal string memory (probably a bit of the time is taken up with memory management at this point, I don't think the initial string table is going to be large enough! or maybe it has to rehash the strings?). After this first loop, you're still concatenating strings, but they're then found in the internal lookup so it doesn't take so long.

That would be my assumption .. 

If you're interested: https://github.com/adobe/avmplus/blob/master/core/AvmCore.cpp#L4231

**Reply by ajwfrost** [reply]
@2jfw the JIT is done function by function, and all this is happening in a single function, so that shouldn't affect things here. But if you split it into different functions and called them repeatedly, yes that could cause a difference, the first function call may take a bit of extra time because of the machine code generation process..

**Reply by 2jfw** [reply]
> I think the "things" are the strings... when you use a string to access a property of an object, it gets internalized, so essentially you've got a million different string concatenations going on, and then each string is set up in the internal string memory (probably a bit of the time is taken up with memory management at this point, I don't think the initial string table is going to be large enough! or maybe it has to rehash the strings?). After this first loop, you're still concatenating strings, but they're then found in the internal lookup so it doesn't take so long.
> 
> That would be my assumption ..

Yes, that seems to be it!
I changed the code/strings to the following ("abc"@first loop, "def"@2nd loop, rather than both using "property") to avoid reusing the same strings and the second loop now has similar timings:

```
// Actual start of test
			time = getTimer();
			for (i = 0; i < len; i++)
			{
				_obj["abc" + i] = i;
			}

			duration = getTimer() - time;


			time2 = getTimer();
			for (j = 0; j < len2; j++)
			{
				_obj2["def" + j] = j;
			}
```

Timings:
7444ms
9877ms
523ms (_obj_ 2nd exectue call)
1555ms (_obj2_ 2nd exectue call)

**Reply by 2jfw** [reply]
Wondering now why the 2nd loop is now _reproducibly_ slower...
6933ms (obj)
9932ms (obj2)
562ms (obj)
1578ms (obj2)

**Reply by ajwfrost** [reply]
Possibly it's just having to rework more... at least at the first time of execution, you're looking at another million strings being interned, and each time the table has to be expanded, it's going to take a bit longer to sort it all out.
For the second time of execution though .. that's a bit odd, it shouldn't really be a factor of the actual strings themselves (although maybe it is, I'm not sure how the hashing works but if all the strings that were added later are then taking longer to find because they're later in the table..?) - slightly odd one there.

What happens if you add a third object/loop and strings `["xyz" + k]`, are these taking even longer?
And what about if the first time round, you do obj and then obj2, but the second time round you do obj2 and then obj?

**Reply by 2jfw** [reply]
[For the sake of completeness] - Changing the number of iterations results in:

**100k loop:**
137ms (obj)
155ms (obj2)
46ms (obj, 2nd call)
59ms (obj2, 2nd call)


**500k loop:**
1009ms (obj)
4076ms (obj2)
245ms (obj, 2nd call)
643ms (obj2, 2nd call)

**Reply by 2jfw** [reply]
> What happens if you add a third object/loop and strings `["xyz" + k]`, are these taking even longer?

Result:

**currentLoop: 1 of 3:**
1079ms (i, _obj)
4130ms (j, _obj2)
3449ms (k, _obj3)

**currentLoop: 2 of 3:**
266ms (i, _obj)
640ms (j, _obj2)
3421ms (k, _obj3)

**currentLoop: 3 of 3:**
268ms (i, _obj)
648ms (j, _obj2)
3396ms (k, _obj3)

Updated Code to:
```

package
{

	import flash.display.Sprite;
	import flash.text.TextField;
	import flash.utils.getTimer;
	import flash.utils.setTimeout;


	public class LoopPerformance extends Sprite
	{

		private var _outputTextField : TextField;
		private var _obj : Object      = {};
		private var _obj2 : Object     = {};
		private var _obj3 : Object     = {};
		private var _numLoops : int    = 3;
		private var _currentLoop : int = 1;


		public function LoopPerformance()
		{
			createOutputTextField();

			setTimeout(execute,
			           1000); // intentionally delayed call to have the window etc get initialized (/ possible AIR internals e.g.)
		}


		private function createOutputTextField() : void
		{
			_outputTextField        = new TextField();
			_outputTextField.width  = 400;
			_outputTextField.height = 300;

			addChild(_outputTextField);
		}


		public function execute() : void
		{
			// Declaring all at the beginning and intentionally not recycling the variables!

			var iterations : int = 500000;

			var i : int;
			var j : int;
			var k : int;

			var len : int  = iterations;
			var len2 : int = iterations;
			var len3 : int = iterations;

			var time : Number;
			var time2 : Number;
			var time3 : Number;

			var duration : Number;
			var duration2 : Number;
			var duration3 : Number;


			// Actual start of test

			/// 1st
			time = getTimer();
			for (i = 0; i < len; i++)
			{
				_obj["abc" + i] = i;
			}
			duration = getTimer() - time;


			/// 2nd
			time2 = getTimer();
			for (j = 0; j < len2; j++)
			{
				_obj2["def" + j] = j;
			}
			duration2 = getTimer() - time2;


			/// 3rd
			time3 = getTimer();
			for (k = 0; k < len3; k++)
			{
				_obj3["ghi" + k] = k;
			}
			duration3 = getTimer() - time3;


			_outputTextField.text += "\n\ncurrentLoop: " + _currentLoop + " of " + _numLoops + ":\n" + duration + "ms (i, _obj)\n" + duration2 + "ms (j, _obj2)\n" + duration3 + "ms (k, _obj3)";

			if (_currentLoop < _numLoops)
			{
				_currentLoop++;

				setTimeout(execute,
				           1000);
			}
		}
	}
}
```

**Reply by 2jfw** [reply]
> And what about if the first time round, you do obj and then obj2, but the second time round you do obj2 and then obj?

Results using inline functions (updates source below and attached)

**Traces:**

calling first
calling second
calling third

currentLoop 1 of 3:
1109ms (i, _obj)
4084ms (j, _obj2)
3874ms (k, _obj3)

=========================

calling second
calling first
calling third

currentLoop 2 of 3:
261ms (i, _obj)
654ms (j, _obj2)
3519ms (k, _obj3)

=========================

calling third
calling second
calling first

currentLoop 3 of 3:
259ms (i, _obj)
663ms (j, _obj2)
3697ms (k, _obj3)

=========================


**Source:**

```
package
{

	import flash.display.Sprite;
	import flash.text.TextField;
	import flash.utils.getTimer;
	import flash.utils.setTimeout;


	public class LoopPerformance extends Sprite
	{

		private var _outputTextField : TextField;
		private var _obj : Object      = {};
		private var _obj2 : Object     = {};
		private var _obj3 : Object     = {};
		private var _numLoops : int    = 3;
		private var _currentLoop : int = 1;


		public function LoopPerformance()
		{
			createOutputTextField();

			setTimeout(execute,
			           1000); // intentionally delayed call to have the window etc get initialized (/ possible AIR internals e.g.)
		}


		private function createOutputTextField() : void
		{
			_outputTextField        = new TextField();
			_outputTextField.width  = 400;
			_outputTextField.height = 1000;

			addChild(_outputTextField);
		}


		public function execute() : void
		{
			// Declaring all at the beginning and intentionally not recycling the variables!

			var iterations : int = 500000;

			var i : int;
			var j : int;
			var k : int;

			var len : int  = iterations;
			var len2 : int = iterations;
			var len3 : int = iterations;

			var time : Number;
			var time2 : Number;
			var time3 : Number;

			var duration : Number;
			var duration2 : Number;
			var duration3 : Number;



			// Actual start of test

			var first : Function = function () : void {
				/// 1st
				_outputTextField.text += "calling first\n";

				time = getTimer();
				for (i = 0; i < len; i++)
				{
					_obj["abc" + i] = i;
				}
				duration = getTimer() - time;
			}


			var second : Function = function () : void {
				/// 2nd
				_outputTextField.text += "calling second\n";

				time2 = getTimer();
				for (j = 0; j < len2; j++)
				{
					_obj2["def" + j] = j;
				}
				duration2 = getTimer() - time2;
			}


			var third : Function = function () : void {
				/// 3rd
				_outputTextField.text += "calling third\n";

				time3 = getTimer();
				for (k = 0; k < len3; k++)
				{
					_obj3["ghi" + k] = k;
				}
				duration3 = getTimer() - time3;
			}


			if (_currentLoop == 1)
			{
				first();
				second();
				third();
			}
			else if (_currentLoop == 2)
			{
				second();
				first();
				third();
			}
			else if (_currentLoop == 3)
			{
				third();
				second();
				first();
			}

			_outputTextField.text += "\ncurrentLoop " + _currentLoop + " of " + _numLoops + ":\n" + duration + "ms (i, _obj)\n" + duration2 + "ms (j, _obj2)\n" + duration3 + "ms (k, _obj3)\n\n=========================\n\n";

			if (_currentLoop < _numLoops)
			{
				_currentLoop++;

				setTimeout(execute,
				           1000);
			}
		}
	}
}
```
**IntelliJ source files without binaries:**
[LoopPerformance v3.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/11219763/LoopPerformance.v3.zip)

**Reply by 2jfw** [reply]
Conclusion here is, that the call order does not seem to matter - the third function (obj3) is always the slowest, even if it is called _first_ (see third overall iteration)

**Reply by 2jfw** [reply]
With 6 loops, it also shows that third function is always the slowest (debug and release build show _no_ significant difference, for the record):

calling first
calling second
calling third

currentLoop 1 of 6:
1055ms (i, _obj)
4164ms (j, _obj2)
3616ms (k, _obj3)

=========================

calling second
calling first
calling third

currentLoop 2 of 6:
274ms (i, _obj)
646ms (j, _obj2)
3412ms (k, _obj3)

=========================

calling third
calling second
calling first

currentLoop 3 of 6:
272ms (i, _obj)
644ms (j, _obj2)
3377ms (k, _obj3)

=========================

calling first
calling second
calling third

currentLoop 4 of 6:
273ms (i, _obj)
646ms (j, _obj2)
3419ms (k, _obj3)

=========================

calling second
calling first
calling third

currentLoop 5 of 6:
273ms (i, _obj)
660ms (j, _obj2)
3481ms (k, _obj3)

=========================

calling third
calling second
calling first

currentLoop 6 of 6:
270ms (i, _obj)
650ms (j, _obj2)
3448ms (k, _obj3)

=========================

**Reply by ajwfrost** [reply]
Yes, so I guess it must be something to do with those strings (`"ghi" + k`) and where they come in the lookup table. I thought they were all hashed which should mean the fact they've got the same first three characters is mostly irrelevant, but potentially it's having to go through all the "abc..." ones and then "def..." ones before it finds these.

I guess it could either be alphabetical-based, or it could be based on when they were added. For academic interest, what happens if you change the "abc" in your first loop to "xyz" to push there first entries into the end of an alphabetical lookup?

And just from a practical standpoint .. I'm assuming there's nothing actually that needs doing here, and it's more about learning how things work internally so that folk are able to avoid pitfalls and to help optimisations?

thanks

**Reply by 2jfw** [reply]
> I guess it could either be alphabetical-based, or it could be based on when they were added. For academic interest, what happens if you change the "abc" in your first loop to "xyz" to push there first entries into the end of an alphabetical lookup?


Sure, here we go with the following result:

500000 iterations loops

calling first with pattern 'xyz'
calling second with pattern 'def'
calling third with pattern 'ghi'

currentLoop 1 of 3:
717ms (i, _obj)
787ms (j, _obj2)
1274ms (k, _obj3)

=========================

500000 iterations loops

calling second with pattern 'def'
calling first with pattern 'xyz'
calling third with pattern 'ghi'

currentLoop 2 of 3:
231ms (i, _obj)
344ms (j, _obj2)
1197ms (k, _obj3)

=========================

500000 iterations loops

calling third with pattern 'ghi'
calling second with pattern 'def'
calling first with pattern 'xyz'

currentLoop 3 of 3:
228ms (i, _obj)
347ms (j, _obj2)
1234ms (k, _obj3)

=========================


> And just from a practical standpoint .. I'm assuming there's nothing actually that needs doing here, and it's more about learning how things work internally so that folk are able to avoid pitfalls and to help optimisations?

To be honest, I actually can hardly tell 🤔 
I would think that this scenario is kind of rare and most users will not encounter it, including the excessive loop count... the impact is likely to be ignored

**Reply by MalacTheLittle** [reply]
> I think the "things" are the strings...

From my test results declaring object properties as Number is 100x to 1000x faster then declaring it with Strings.
But still "strings" doesn't explain time difference between each loop since every property name is unique in all loops...

Each time time difference increases from time_13 till time_17, then it drops... 

```
var object:Object = {};
setTimeout(execute, 500, object);

function execute(object:Object):void{
	for(var nr:uint = 1; nr < 21; ++nr){
		test(nr, object);
	}
}

function test(nr:uint, object:Object):void{
	var len:Number = 50000;
	var si:uint = nr*len;
	var ei:uint = si + len;
	
	var startTime:Number = getTimer();
	for(var i:uint = si; i < ei; i++){
		object["prop"+i] = i;
	}
	trace("time_" + nr + ", 	adding prop" + si + " to prop" + ei + ":	" + (getTimer()-startTime), "ms");
}
```
```
OUTPUT:
time_1, 	adding prop50000 to prop100000:	130 ms
time_2, 	adding prop100000 to prop150000:	92 ms
time_3, 	adding prop150000 to prop200000:	91 ms
time_4, 	adding prop200000 to prop250000:	172 ms
time_5, 	adding prop250000 to prop300000:	111 ms
time_6, 	adding prop300000 to prop350000:	128 ms
time_7, 	adding prop350000 to prop400000:	227 ms
time_8, 	adding prop400000 to prop450000:	394 ms
time_9, 	adding prop450000 to prop500000:	210 ms
time_10, 	adding prop500000 to prop550000:	159 ms
time_11, 	adding prop550000 to prop600000:	248 ms
time_12, 	adding prop600000 to prop650000:	227 ms
time_13, 	adding prop650000 to prop700000:	541 ms
time_14, 	adding prop700000 to prop750000:	1275 ms
time_15, 	adding prop750000 to prop800000:	1149 ms
time_16, 	adding prop800000 to prop850000:	2422 ms
time_17, 	adding prop850000 to prop900000:	1657 ms
time_18, 	adding prop900000 to prop950000:	171 ms
time_19, 	adding prop950000 to prop1000000:	155 ms
time_20, 	adding prop1000000 to prop1050000:	32 ms
```

**Reply by ajwfrost** [reply]
Curious that it drops down again.. but I still think this is something to do with the interned string table. If you use a number, it does get converted to a string, but probably a bit quicker than you'd be able to do with the extra concatenation step here. Probably when the table is getting fuller there's a point at which it's completely revised and then perhaps there's more space which makes the subsequent adds a little quicker.

Anyone who wants to grab the avmplus code and build avmshell, and then add some internal profiling/debugging, to work out exactly what's happening, please feel free :-)

thanks

**Comment by 2jfw** [comment]
BTW, to give a short scope why I did actually initially create the test was to find out if assigning a class member to a function scope variable and then accessing this one in a loop would be faster - I changed the test to avoid the String "issue" and can confirm that this is actually **NOT** the case.
I somehow had a test of Jackson Dunstan in mind where it was proven that accessing class members is slower than accessing scope(d) variables...

Code-wise:

```
// Class Member access
time3 = getTimer();
for (k = 0; k < len3; k++)
{
   _obj3[pattern] = k;
}
duration3 = getTimer() - time3;

==

// Scope Member access
time3 = getTimer();
var scopeObj : Object = _obj3;
for (k = 0; k < len3; k++)
{
   scopeObj[pattern] = k;
}
duration3 = getTimer() - time3;
```

The timings are the same and there is no difference.

**Hint:**
Using a `Getter` for _obj3 in a loop is (..oh wonder!) **way** slower and you all should definitely avoid this and cache the value of the `Getter` to a scope variable before looping (**EDIT:** this does of course only apply to the scenario where the value of the `Getter` would not change whilst iterating through the loop)
